Add Collation-Based String Sorting

Drodt opened the pull request:
Previously, Hayagriva did not compare UTF-8 strings correctly #193. This is due to how Rust compares UTF-8 Strings. This PR addresses the issue by using `icu_ucol` for comparison, using the current locale.

Fixes #193 and adds one test to citeproc-pass.

Adds `rust_icu_ucol` as a dependency. Problem: icu must be installed on the user's system. How do we handle this?

I might have missed some comparison in `csl_cmp`.

DerDrodt committed:
Add collation cmp

DerDrodt committed:
Add local test for #193

DerDrodt committed:
Clippy

Drodt commented:
BTW: This is only a draft because I don't think it should be merged without discussing the added dependency.

Any opinion on this, @reknih @PgBiel?

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Merge branch 'main' into collation

PgBiel commented:
> Problem: icu must be installed on the user's system. How do we handle this?

It's a non-starter for us. It won't be installed on the Typst web app for example.

icu4x[^1] is an alternative. However, there are concerns about the size of the collation data it pulls, and how this could impact the size of e.g. the Typst binary, Typst on the web app etc.

One alternative that was considered is to delegate this job to a trait which implements localized sorting and let upstream users (i.e. Typst) handle it. We're not fully decided on this, but it's one route to consider.

[^1]: https://github.com/unicode-org/icu4x

DerDrodt committed:
Switch to icu4x

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Update archives

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Update Style

PgBiel committed:
further update

DerDrodt committed:
Merge branch 'main' into update-archives

DerDrodt committed:
Update archives
- CSE renamed
- mhra renamed
- mla 8th edition removed
- vancouver renamed to nlm

DerDrodt committed:
Update archives

DerDrodt committed:
Fix test (add locales)

Drodt committed:
Merge pull request #453 from DerDrodt/update-archives

Update archives

DerDrodt committed:
Merge branch 'main' into collation

DerDrodt committed:
Merge remote-tracking branch 'upstream/main' into collation

DerDrodt committed:
Fix merge error

DerDrodt committed:
Fix icu4x version 2.2.0

laurmaedje commented:
Turns out that this adds quite a bit of binary size (~1.2MB) to Typst and still ~400KB when brotli compressed (for our web app). This is a bit unfortunate, but not much we can do right now. But it intensifies the need for some sort of lazy data loading in Typst and hayagriva. For the CLI, it's not that important, but the web app wasm blob keeps growing (from 7.1MB brotli for 0.14.2 to 7.7MB brotli for the latest version including the to-be-done hayagriva bump).
